internal/app/server/server.go

```diff
@@ -0,0 +1,54 @@
+// Copyright 2020 Envoyproxy Authors
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package server
+
+import (
+	"context"
+	"net"
+
+	"github.com/envoyproxy/xds-relay/internal/pkg/log"
+
+	api "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"github.com/envoyproxy/go-control-plane/pkg/cache"
+	gcp "github.com/envoyproxy/go-control-plane/pkg/server"
+	"google.golang.org/grpc"
+)
+
+// Run instantiates a running gRPC server for accepting incoming xDS-based
+// requests.
+func Run() {
+	logger := log.New().Sugar()
+
+	// Cursory implementation of go-control-plane's server.
+	snapshotCache := cache.NewSnapshotCache(false, cache.IDHash{}, nil)
+	gcpServer := gcp.NewServer(context.Background(), snapshotCache, nil)
+	server := grpc.NewServer()
+	listener, err := net.Listen("tcp", ":8080")
+	if err != nil {
+		logger.Fatalw("failed to bind server to listener", "err", err)
+	}
+
+	api.RegisterEndpointDiscoveryServiceServer(server, gcpServer)
+	api.RegisterClusterDiscoveryServiceServer(server, gcpServer)
+	api.RegisterRouteDiscoveryServiceServer(server, gcpServer)
+	api.RegisterListenerDiscoveryServiceServer(server, gcpServer)
+
+	logger.Info("Initializing server at", listener.Addr().String())
+	go func() {
+		if err := server.Serve(listener); err != nil {
+			logger.Fatalw("failed to initialize server", "err", err)
+		}
+	}()
```

Comment:
does the main loop exit and so does the process. We have to add some kind of `waitGroup.wait()` logic here.

Reply:
Yes, let me take the server out of the Go routine and serve this on the main thread.

Reply:
or may be leave the server out in the initial PR to unblock other streams? We do need to add some other aspects to serve also like graceful shutdown as well at a later point.

Reply:
It shouldn't block other streams, the cache / mapper / control plane client are primarily interfaces. We can always serve additional non-blocking workflows on separate Go routines. I think the server should run on the main thread, but let me know if I missed something.

Reply:
ah..i meant the workstreams to parallelize the effort once the scaffolding is in.